openedx_learning/apps/authoring/publishing/models/draft_log.py

```diff
@@ -0,0 +1,239 @@
+"""
+Models relating to the creation and management of Draft information.
+"""
+from django.conf import settings
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+
+from openedx_learning.lib.fields import immutable_uuid_field, manual_date_time_field
+
+from .learning_package import LearningPackage
+from .publishable_entity import PublishableEntity, PublishableEntityVersion
+
+
+class Draft(models.Model):
+    """
+    Find the active draft version of an entity (usually most recently created).
+
+    This model mostly only exists to allow us to join against a bunch of
+    PublishableEntity objects at once and get all their latest drafts. You might
+    use this together with Published in order to see which Drafts haven't been
+    published yet.
+
+    A Draft entry should be created whenever a new PublishableEntityVersion is
+    created. This means there are three possible states:
+
+    1. No Draft entry for a PublishableEntity: This means a PublishableEntity
+       was created, but no PublishableEntityVersion was ever made for it, so
+       there was never a Draft version.
+    2. A Draft entry exists and points to a PublishableEntityVersion: This is
+       the most common state.
+    3. A Draft entry exists and points to a null version: This means a version
+       used to be the draft, but it's been functionally "deleted". The versions
+       still exist in our history, but we're done using it.
+
+    It would have saved a little space to add this data to the Published model
+    (and possibly call the combined model something else). Split Modulestore did
+    this with its active_versions table. I keep it separate here to get a better
+    separation of lifecycle events: i.e. this table *only* changes when drafts
+    are updated, not when publishing happens. The Published model only changes
+    when something is published.
+    """
+    # If we're removing a PublishableEntity entirely, also remove the Draft
+    # entry for it. This isn't a normal operation, but can happen if you're
+    # deleting an entire LearningPackage.
+    entity = models.OneToOneField(
+        PublishableEntity,
+        on_delete=models.CASCADE,
+        primary_key=True,
+    )
+    version = models.OneToOneField(
+        PublishableEntityVersion,
+        on_delete=models.RESTRICT,
+        null=True,
+        blank=True,
+    )
+
+
+class DraftChangeLog(models.Model):
+    """
+    There is one row in this table for every time Drafts are created/modified.
+
+    Most of the time we'll only be changing one Draft at a time, and this will
+    be 1:1 with DraftChange. But there are some operations that affect many
```

Comment:
Would such "some operations" include cases where parent containers are changed via a DraftSideEffect? If so, won't that actually be most operations, since most PEs have a parent container?

Reply:
I guess I was thinking more of present-state libraries when writing this. I'll change it.